segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/GarbageCollector.java

```diff
@@ -0,0 +1,385 @@
+        val chunksToDelete = new ArrayList<GarbageChunkInfo>();
+        int count = 0;
+        try {
+            // Block until you have at least one item.
+            GarbageChunkInfo info = garbageChunks.take();
+            log.trace("{}: deleteGarbage - retrieved {}", traceObjectId, info);
+            while (null != info ) {
+                queueSize.decrementAndGet();
+                chunksToDelete.add(info);
+
+                count++;
+                if (count >= maxItems) {
+                    break;
+                }
+                // Do not block
+                info = garbageChunks.poll();
+                log.trace("{}: deleteGarbage - retrieved {}", traceObjectId, info);
+            }
+        } catch (InterruptedException e) {
+            throw new CompletionException(e);
+        }
+
+        // Sleep if no chunks to delete.
+        if (count == 0) {
+            log.debug("{}: deleteGarbage - no work - sleeping for {}.", traceObjectId, config.getGarbageCollectionDelay());
+            return CompletableFuture.completedFuture(false);
+        }
+
+        // For each chunk delete if the chunk is not present at all in the metadata or is present but marked as inactive.
+        ArrayList<CompletableFuture<Void>> futures = new ArrayList<>();
+        for (val infoToDelete : chunksToDelete) {
+            val chunkToDelete = infoToDelete.name;
+            val failed = new AtomicBoolean();
+            val txn = metadataStore.beginTransaction(false, chunkToDelete);
+            val future =
+                    txn.get(infoToDelete.name)
+                            .thenComposeAsync(metadata -> {
+                                val chunkMetadata = (ChunkMetadata) metadata;
+                                // Delete if the chunk is not present at all in the metadata or is present but marked as inactive.
+                                val shouldDeleteChunk = null == chunkMetadata || !chunkMetadata.isActive();
+                                val shouldDeleteMetadata = new AtomicBoolean(null != metadata && !chunkMetadata.isActive());
+
+                                // Delete chunk from storage.
+                                if (shouldDeleteChunk) {
+                                    return chunkStorage.openWrite(chunkToDelete)
+                                            .thenComposeAsync(chunkStorage::delete, storageExecutor)
+                                            .handleAsync((v, e) -> {
+                                                if (e != null) {
+                                                    val ex = Exceptions.unwrap(e);
+                                                    if (ex instanceof ChunkNotFoundException) {
+                                                        // Ignore - nothing to do here.
+                                                        log.debug("{}: deleteGarbage - Could not delete garbage chunk={}.", traceObjectId, chunkToDelete);
```

Comment:
Maybe is a personal impression, but in general I feel that the code is heavily influenced by logging. There are many `if/else` branches, computation stages/futures, etc. that exist just to add debug logs. I don't know if this is related to the need of explicitly stating what the program is doing at any time, as there are scenarios that we are unsure of its behavior. If this is the case, maybe we could remove logging by adding more tests and splitting large code blocks into smaller functions that can be heavily tested. Maybe this would give us more certainty and this, in turn, would help us to remove some of the log lines in the PR. But, even in the case that we decide to do not do that now, I wonder if so much logging would be counterproductive, as logs will grow significantly, making them harder to manage.

Reply:
There are 5 info level logs which I find important
```
log.info ("{}: Iteration {} started.", traceObjectId, iterationId.get());
log.info("{}: Iteration {} ended.", traceObjectId, iterationId.getAndIncrement());
log.info("{}: deleteGarbage - suspended - sleeping for {}.", traceObjectId, config.getGarbageCollectionDelay());
log.info("{}: deleteGarbage - Chunk is not marked as garbage chunk={}.", traceObjectId, chunkToDelete);
log.info("{}: deleteGarbage - could not delete after max attempts chunk={}.", traceObjectId, chunkToDelete);
```

Following are important to know what went wrong
```
log.warn("{}: deleteGarbage - Queue full. Could not delete garbage. Chunks skipped", traceObjectId);
log.warn("{}: deleteGarbage - Could not delete garbage chunk={}.", traceObjectId, chunkToDelete);
                                                        
```

following are unusual circumstances that need to be recorded

```
log.error("{}: Error during doRun.", traceObjectId, ex);
log.error(String.format("%s deleteGarbage - Could not delete metadata for garbage chunk=%s.",
                                                            traceObjectId, chunkToDelete), e);
log.error(String.format("%s deleteGarbage - Could not find garbage chunk=%s.",
                                            traceObjectId, chunkToDelete), ex);
```

Perhaps last two above could be info or warn.


Other logs are for debug purpose and as such they need to show details of what happened ....
we'll tweak this as we go along ..some of them will become trace as we gain more run time with this code.